README.md
[![Build Status](https://travis-ci.org/fivdi/onoff.svg?branch=master)](https://travis-ci.org/fivdi/onoff)
[![codecov](https://codecov.io/gh/fivdi/onoff/branch/master/graph/badge.svg)](https://codecov.io/gh/fivdi/onoff)
[![npm Version](http://img.shields.io/npm/v/onoff.svg)](https://www.npmjs.com/package/onoff)
[![Downloads Per Month](http://img.shields.io/npm/dm/onoff.svg)](https://www.npmjs.com/package/onoff)
[![Mentioned in Awesome Node.js](https://awesome.re/mentioned-badge.svg)](https://github.com/sindresorhus/awesome-nodejs#hardware)

# onoff

GPIO access and interrupt detection with **Node.js** on Linux boards like the
Raspberry Pi or BeagleBone.

onoff supports Node.js versions 10, 12, 14, 15 and 16.

## Contents

 * [Installation](#installation)
 * [Usage](#usage)
   * [LEDs and Buttons](#leds-and-buttons)
   * [Debouncing Buttons](#debouncing-buttons)
   * [Blink an LED Using the Synchronous API](#blink-an-led-using-the-synchronous-api)
   * [Blink an LED Using the Asynchronous API and Completion Callbacks](#blink-an-led-using-the-asynchronous-api-and-completion-callbacks)
   * [Blink an LED Using the Asynchronous API and Promises](#blink-an-led-using-the-asynchronous-api-and-promises)
 * [API](#api)
 * [How Does onoff Work?](#how-does-onoff-work)
 * [Configuring Pullup and Pulldown Resistors](#configuring-pullup-and-pulldown-resistors)
 * [Benchmarks](#benchmarks)
 * [Related Packages](#related-packages)
 * [Additional Information](#additional-information)

## Installation

```
npm install onoff
```

Note that although it's possible to install onoff on non-Linux systems the
functionality offered by onoff is only available on Linux systems.

## Usage

#### LEDs and Buttons
Assume that there's an LED connected to GPIO17 and a momentary push button
connected to GPIO4.

<img src="https://raw.githubusercontent.com/fivdi/onoff/master/examples/light-switch.png">

When the button is pressed the LED should turn on, when it's released the LED
should turn off. This can be achieved with the following code:

```js
const Gpio = require('onoff').Gpio;
const led = new Gpio(17, 'out');
const button = new Gpio(4, 'in', 'both');

button.watch((err, value) => led.writeSync(value));
```

Here two Gpio objects are being created. One called led for the LED connected
to GPIO17 which is an output, and one called button for the momentary push
button connected to GPIO4 which is an input. In addition to specifying that
the button is an input, the constructors optional third argument is used to
specify that 'both' rising and falling interrupt edges should be configured
for the button GPIO as both button presses and releases should be handled.

After everything has been setup correctly, the buttons watch method is used to
specify a callback function to execute every time the button is pressed or
released. The value argument passed to the callback function represents the
state of the button which will be 1 for pressed and 0 for released. This value
is used by the callback to turn the LED on or off using its writeSync method.

When the above program is running it can be terminated with ctrl-c. However,
it doesn't free its resources. It also ignores the err argument passed to
the callback. Here's a slightly modified variant of the program that handles
ctrl-c gracefully and bails out on error. The resources used by the led and
button Gpio objects are released by invoking their unexport method.

```js
const Gpio = require('onoff').Gpio;
const led = new Gpio(17, 'out');
const button = new Gpio(4, 'in', 'both');

button.watch((err, value) => {
  if (err) {
    throw err;
  }

  led.writeSync(value);
});

process.on('SIGINT', _ => {
  led.unexport();
  button.unexport();
});
```

#### Debouncing Buttons
When working with buttons there will often be button bounce issues which
result in the hardware thinking that a button was pressed several times
although it was only pressed once. onoff provides a software debouncing
solution for resolving bounce issues.

Assume again that there's an LED connected to GPIO17 and a momentary push
button connected to GPIO4.

When the button is pressed the LED should toggle its state. This is a typical
example of a situation where there will be button bounce issues. The issue can
be resolved by using the debounceTimeout option when creating the Gpio object
for the button. In the below program the debounceTimeout is set to 10
milliseconds. This delays invoking the watch callback for the button while the
button is bouncing. The watch callback will not be invoked until the button
stops bouncing and has been in a stable state for 10 milliseconds.

```js
const Gpio = require('onoff').Gpio;
const led = new Gpio(17, 'out');
const button = new Gpio(4, 'in', 'rising', {debounceTimeout: 10});

button.watch((err, value) => {
  if (err) {
    throw err;
  }

  led.writeSync(led.readSync() ^ 1);
});

process.on('SIGINT', _ => {
  led.unexport();
  button.unexport();
});
```

#### Blink an LED Using the Synchronous API

Blink an LED connected to GPIO17 for 5 seconds using the synchronous readSync
and writeSync methods.

```js
const Gpio = require('../onoff').Gpio; // Gpio class
const led = new Gpio(17, 'out');       // Export GPIO17 as an output

// Toggle the state of the LED connected to GPIO17 every 200ms
const iv = setInterval(_ => led.writeSync(led.readSync() ^ 1), 200);

// Stop blinking the LED after 5 seconds
setTimeout(_ => {
  clearInterval(iv); // Stop blinking
  led.unexport();    // Unexport GPIO and free resources
}, 5000);
```

#### Blink an LED Using the Asynchronous API and Completion Callbacks

Blink an LED connected to GPIO17 for 5 seconds using the asynchronous read and
write methods and completion callbacks.

```js
const Gpio = require('../onoff').Gpio; // Gpio class
const led = new Gpio(17, 'out');       // Export GPIO17 as an output
let stopBlinking = false;

// Toggle the state of the LED connected to GPIO17 every 200ms
const blinkLed = _ => {
  if (stopBlinking) {
    return led.unexport();
  }

  led.read((err, value) => { // Asynchronous read
    if (err) {
      throw err;
    }

    led.write(value ^ 1, err => { // Asynchronous write
      if (err) {
        throw err;
      }
    });
  });

  setTimeout(blinkLed, 200);
};

blinkLed();

// Stop blinking the LED after 5 seconds
setTimeout(_ => stopBlinking = true, 5000);
```

#### Blink an LED Using the Asynchronous API and Promises

Blink an LED connected to GPIO17 for 5 seconds using the asynchronous read and
write methods and Promises.

```js
const Gpio = require('../onoff').Gpio; // Gpio class
const led = new Gpio(17, 'out');       // Export GPIO17 as an output
let stopBlinking = false;

// Toggle the state of the LED connected to GPIO17 every 200ms
const blinkLed = _ => {
  if (stopBlinking) {
    return led.unexport();
  }

  led.read()
    .then(value => led.write(value ^ 1))
    .then(_ => setTimeout(blinkLed, 200))
    .catch(err => console.log(err));
};

blinkLed();

// Stop blinking the LED after 5 seconds
setTimeout(_ => stopBlinking = true, 5000);
```

#### Check accessibility

Sometimes it may be necessary to determine if the current system supports
GPIOs programmatically and mock functionality if it doesn't. `Gpio.accessible`
can be used to achieve this.

```js
const Gpio = require('onoff').Gpio;

const useLed = (led, value) => led.writeSync(value);

let led;

if (Gpio.accessible) {
  led = new Gpio(17, 'out');
  // more real code here
} else {
  led = {
    writeSync: value => {
      console.log('virtual led now uses value: ' + value);
    }
  };
}

useLed(led, 1);
```

## API

### Class Gpio

  * [Gpio(gpio, direction [, edge] [, options]) - Constructor](#gpiogpio-direction--edge--options)
  * [read([callback]) - Read GPIO value asynchronously](#readcallback)
  * [readSync() - Read GPIO value synchronously](#readsync)
  * [write(value[, callback]) - Write GPIO value asynchronously](#writevalue-callback)
  * [writeSync(value) - Write GPIO value synchronously](#writesyncvalue)
  * [watch(callback) - Watch for hardware interrupts on the GPIO](#watchcallback)
  * [unwatch([callback]) - Stop watching for hardware interrupts on the GPIO](#unwatchcallback)
  * [unwatchAll() - Remove all watchers for the GPIO](#unwatchall)
  * [direction() - Get GPIO direction](#direction)
  * [setDirection(direction) - Set GPIO direction](#setdirectiondirection)
  * [edge() - Get GPIO interrupt generating edge](#edge)
  * [setEdge(edge) - Set GPIO interrupt generating edge](#setedgeedge)
  * [activeLow() - Get GPIO activeLow setting](#activelow)
  * [setActiveLow(invert) - Set GPIO activeLow setting](#setactivelowinvert)
  * [unexport() - Reverse the effect of exporting the GPIO to userspace](#unexport)
  * [static accessible - Determine whether or not GPIO access is possible](#static-accessible)
  * [HIGH / LOW - Constants used when reading or writing a GPIO value](#static-high--low)

##### Gpio(gpio, direction [, edge] [, options])
- gpio - An unsigned integer specifying the GPIO number.
- direction - A string specifying whether the GPIO should be configured as an
input or output. The valid values are: 'in', 'out', 'high', and 'low'. If 'out'
is specified the GPIO will be configured as an output and the value of the GPIO
will be set to 0. 'high' and 'low' are variants of 'out' that configure the
GPIO as an output with an initial level of 1 or 0 respectively.
- [edge] - An optional string specifying the interrupt generating edge or
edges for an input GPIO. The valid values are: 'none', 'rising', 'falling' or
'both'. The default value is 'none' indicating that the GPIO will not generate
interrupts. Whether or not interrupts are supported by an input GPIO is GPIO
specific. If interrupts are not supported by a GPIO the edge argument should
not be specified. The edge argument is ignored for output GPIOs.
- [options] - An optional options object.

Configures the GPIO based on the passed arguments and returns a new Gpio
object that can be used to access the GPIO.

The following options are supported:
- debounceTimeout - An unsigned integer specifying a millisecond delay. Delays
invoking the watch callback for an interrupt generating input GPIO while the
input is bouncing. The watch callback will not be invoked until the input
stops bouncing and has been in a stable state for debounceTimeout
milliseconds. Optional, if unspecified the input GPIO will not be debounced.
- activeLow - A boolean value specifying whether the values read from or
written to the GPIO should be inverted. The interrupt generating edge for the
GPIO also follow this this setting. The valid values for activeLow are true
and false. Setting activeLow to true inverts. Optional, the default value is
false.
- reconfigureDirection - A boolean value specifying whether the direction for
the GPIO should be reconfigured even though the direction is already
configured correctly. When an application starts, the direction of a GPIO used
by that application may already be configured correctly, for example, from a
previous run of the application. Reconfiguring the direction of that GPIO can
result in unwanted side effects. For example, if a GPIO is already configured
as an output and it is reconfigured as an output by passing 'out' to the
constructor, the value of that output will be set to 0. In some applications
this is not desirable and the value of the output should not be modified. The
reconfigureDirection option can help here. If reconfigureDirection is set to
false the direction of a GPIO that is already correctly configured will not be
reconfigured. Optional, the default value is true.

GPIOs on Linux are identified by unsigned integers. These are the numbers that
should be passed to the onoff Gpio constructor when exporting GPIOs to
userspace. For example, pin 11 on the Raspberry Pi expansion header
corresponds to GPIO17 in Raspbian Linux. 17 is therefore the number to pass
to the onoff Gpio constructor when using pin 11 on the expansion header.

##### read([callback])
- [callback] - An optional completion callback that gets two arguments (err,
value), where err is reserved for an Error object and value is the number 0
or 1 and represents the state of the GPIO.

Read GPIO value asynchronously. If no completion callback is specified read
returns a Promise which resolves to the value of the GPIO on success or rejects
with an Error object on failure.

Note that most systems support readback of GPIOs configured as outputs. The
read method can therefore be invoked for any GPIO, irrespective of whether it
was configured as an input or an output. The Raspberry Pi and BeagleBone are
examples of such systems.

##### readSync()
Read GPIO value synchronously. Returns the number 0 or 1 to represent the
state of the GPIO.

Note that most systems support readback of GPIOs configured as outputs. The
readSync method can therefore be invoked for any GPIO, irrespective of whether
it was configured as an input or an output. The Raspberry Pi and BeagleBone
are examples of such systems.

##### write(value[, callback])
- value - The number 0 or 1.
- [callback] - An optional completion callback that gets one argument (err),
where err is reserved for an error object.

Write GPIO value asynchronously. If no completion callback is specified write
returns a Promise that resolves with no value on success or rejects with an
Error object on failure.

Note that on most systems invoking write for a GPIO configured as an input
will result in an EPERM error indicating that the operation is not permitted.
The Raspberry Pi and BeagleBone are examples of such systems.

##### writeSync(value)
- value - The number 0 or 1.

Write GPIO value synchronously.

Note that on most systems invoking writeSync for a GPIO configured as an input
will result in an EPERM error indicating that the operation is not permitted.
The Raspberry Pi and BeagleBone are examples of such systems.

##### watch(callback)
- callback - A callback that gets two arguments (err, value), where err is
reserved for an error object and value is the number 0 or 1 and represents the
state of the GPIO. The value can also be used to determine whether the
interrupt occurred on a rising or falling edge. A value of 0 implies a falling
edge interrupt and a value of 1 implies a rising edge interrupt.

Watch for hardware interrupts on the GPIO. The edge argument that was passed
to the constructor determines which hardware interrupts to watch for.

##### unwatch([callback])
- [callback] - The callback to remove.

Stop watching for hardware interrupts on the GPIO. If callback is specified,
only that particular callback is removed. Otherwise all callbacks are removed.

##### unwatchAll()
Remove all hardware interrupt watchers for the GPIO.

##### direction()
Returns the string 'in' or 'out' indicating whether the GPIO is an input or
output.

##### setDirection(direction)
- direction - A string specifying whether the GPIO should be configured as an
input or output. The valid values are 'in', 'out', 'high', and 'low'. If 'out'
is specified the GPIO will be configured as an output and the value of the GPIO
will be set to 0. 'high' and 'low' are variants of 'out' that configure the
GPIO as an output with an initial level of 1 or 0 respectively.

Set GPIO direction.

##### edge()
Returns the string 'none', 'falling', 'rising', or 'both' indicating the
interrupt generating edge or edges for the GPIO. Whether or not interrupts are
supported by an input GPIO is GPIO specific. If interrupts are not supported
the edge method should not be used. Interrupts are not supported by output
GPIOs.

##### setEdge(edge)
- edge - A string specifying the interrupt generating edge or edges for an
input GPIO. The valid values are: 'none', 'rising', 'falling' or 'both'.
Whether or not interrupts are supported by an input GPIO is GPIO specific. If
interrupts are not supported the setEdge method should not be used. Interrupts
are not supported by output GPIOs.

Set GPIO interrupt generating edge.

##### activeLow()
Returns true or false indicating whether or not the values read from or written
to the GPIO are inverted.

##### setActiveLow(invert)
- invert - A boolean value specifying whether the values read from or written
to the GPIO should be inverted. The interrupt generating edge for the GPIO also
follow this this setting. The valid values for invert are true and false.
Setting activeLow to true inverts.

Set GPIO activeLow setting.

##### unexport()
Reverse the effect of exporting the GPIO to userspace. A Gpio object should not
be used after invoking its unexport method.

##### static accessible
Determine whether or not GPIO access is possible. true if the current process
has the permissions required to export GPIOs to userspace. false otherwise.
Loosely speaking, if this property is true it should be possible for the
current process to create Gpio objects.

It is notable that while this property may be false indicating that the
current process does not have the permissions required to export GPIOs to
userspace, existing exported GPIOs may still be accessible.

This property is useful for mocking functionality on computers used for
development that do not provide access to GPIOs.

This is a static property and should be accessed as `Gpio.accessible`.

##### static HIGH / LOW
Constants used when reading or writing a GPIO value. Gpio.HIGH and Gpio.LOW
can be used in place of the numeric constants 1 and 0.


## How Does onoff Work?

Internally onoff uses sysfs files located at /sys/class/gpio to access GPIOs
and the [epoll package](https://github.com/fivdi/epoll) to detect hardware
interrupts. The Linux GPIO sysfs interface for userspace is documented
[here](https://www.kernel.org/doc/Documentation/gpio/sysfs.txt).
It's a relatively simple interface which can be used to ask the Linux kernel
to export control of a GPIO to userspace. After control of a GPIO has been
exported to userspace, the GPIO can be configured as an input or output.
Thereafter, the state of an input can be read, and the state of an output can
be written. Some systems will also allow the state of a output to be read.
The GPIO sysfs interface can also be used for interrupt detection. onoff can
detect several thousand interrupts per second on both the BeagleBone and the
Raspberry Pi.


## Configuring Pullup and Pulldown Resistors

As mentioned in section [How Does onoff Work](#how-does-onoff-work) the sysfs
interface is used to access GPIOs. The sysfs interface doesn't offer support
for configuring pullup and pulldown resistors on GPIOs.

There are however many platform specific mechanisms for configuring pullup and
pulldown resistors that are compatible with onoff. onoff itself doesn't use
these mechanisms as one of the goals of onoff is to be platform independent.

Here we'll take a look at two mechanisms available on the Raspberry Pi for
configuring pullup and pulldown resistors.

The first point to be aware of is that most GPIOs on a Raspberry Pi have
either their pullup or pulldown resistor activated by default. The defaults
can be seen in Table 6-31 on pages 102 and 103 of the
[BCM2835 ARM Peripherals](http://www.farnell.com/datasheets/1521578.pdf)
documentation.

#### Using the gpio Command in /boot/config.txt 

On Raspbian 2018-04-18 or later the `gpio` configuration command can be used
in `/boot/config.txt` to configure pullup and pulldown resistors. Further
information is available at
[New "gpio" config command](https://www.raspberrypi.org/forums/viewtopic.php?f=117&t=208748).

#### Using Device Tree Overlays

Device tree overlays can also be used to configure pullup and pulldown
resistors. The Wiki page
[Enabling Pullup and Pulldown Resistors on The Raspberry Pi](https://github.com/fivdi/onoff/wiki/Enabling-Pullup-and-Pulldown-Resistors-on-The-Raspberry-Pi)
describes this mechanism in more detail.

## Benchmarks

Three of the onoff tests are used to monitor performance.

  * performance-async.js - determine max. no. of write ops per seconds
  * performance-sync.js - determine max. no. of writeSync ops per second
  * performance-interrupt.js - determine max. no. of interrupts per second

The results of these tests are shown in the following tables.

**Raspberry Pi 4 B, 1.5GHz, Raspberry Pi OS (March 4th 2021, Buster 10.8)**

node | onoff | kernel | write / sec | writeSync / sec | interrupts / sec
:---: | :---: | :---: | ---: | ---: | ---:
v16.0.0 | v6.0.3 | 5.10.17-v7l+ | 25124 | 280417 | 20240
v15.14.0 | v6.0.3 | 5.10.17-v7l+ | 24055 | 271149 | 20488
v14.16.1 | v6.0.3 | 5.10.17-v7l+ | 21669 | 254705 | 19703
v12.22.1 | v6.0.3 | 5.10.17-v7l+ | 22618 | 318417 | 21122
v10.24.1 | v6.0.3 | 5.10.17-v7l+ | 22405 | 329927 | 19583

**Raspberry Pi 3 B, 1.2GHz, Raspbian Buster 10.1**

node | onoff | kernel | write / sec | writeSync / sec | interrupts / sec
:---: | :---: | :---: | ---: | ---: | ---:

History.md
6.0.3 - Apr 26 2021
===================

  * update performance numbers
  * add support for node.js 16
  * update dependencies

6.0.2 - Apr 07 2021
===================

  * fix accessible property (thank you [@mildsunrise](https://github.com/mildsunrise))
  * update dependencies
  * add support for node.js 15

6.0.1 - Oct 10 2020
===================

  * switch from coveralls to codecov
  * update dependencies
  * drop support for node.js 8
  * drop support for node.js 13

6.0.0 - Apr 23 2020
===================

  * document the potential of EPERM errors when invoking write methods (fixes [#167](https://github.com/fivdi/onoff/issues/167))
  * drop support for node.js 6, add support for node.js 14
  * avoid calling fs.writeFileSync with numeric data (fixes [#170](https://github.com/fivdi/onoff/issues/170))
  * update dependencies
  * use g++-6 on travis

5.0.1 - Dec 24 2019
===================

  * document node 11 support
  * update benchmark results for pi 1, 2, 3 and 4

5.0.0 - Sep 22 2019
===================

  * drop support for node.js v4
  * update dependencies (epoll v3.0.0, ts-node v8.4.1, typescript v3.6.3)

4.1.4 - Sep 07 2019
===================

  * update dependencies (epoll v2.0.10, coveralls v3.0.6, mocha v6.2.0, typescript v3.6.2)

4.1.3 - Jul 05 2019
===================

  * avoid recursion in read and write methods (fixes [#156](https://github.com/fivdi/onoff/issues/156))

4.1.2 - Jun 16 2019
===================

  * fix export
  * refactor promises (thank you [@pizzaisdavid](https://github.com/pizzaisdavid))
  * update npm keywords
  * update dependencies

4.1.1 - Mar 14 2019
===================

  * simplify constructor
  * update dependencies (epoll v2.0.9, jshint v2.10.2, ts-node v8.0.3)

4.1.0 - Mar 03 2019
===================

  * add type definitions for TypeScript (thank you [@saenglert](https://github.com/saenglert))

4.0.0 - Feb 28 2019
===================

  * added Promises to async read/write operations (thank you [@saenglert](https://github.com/saenglert)) - breaking change 
  * update dependencies (mocha@6.0.2, nyc@13.3.0)

3.2.9 - Feb 24 2019
===================

  * post lcov to coveralls.io

3.2.8 - Feb 21 2019
===================

  * prevent EACCES errors from occurring while waiting for file access permission [#131](https://github.com/fivdi/onoff/issues/131)

3.2.7 - Feb 17 2019
===================

  * add code coverage to build
  * add more unit tests
  * document node 11 support
  * only reconfigure direction if needed [#128](https://github.com/fivdi/onoff/issues/128)

3.2.6 - Feb 09 2019
===================

  * add travis build

3.2.5 - Feb 09 2019
===================

  * lint with jshint

3.2.4 - Feb 09 2019
===================

  * add .npmignore

3.2.3 - Feb 09 2019
===================

  * update dependencies

3.2.2 - Sep 30 2018
===================

  * add unittests for reading and writing (thank you [@pizzaisdavid](https://github.com/pizzaisdavid))
  * update dependencies (epoll v2.0.4, mocha v4.7.0)

3.2.1 - Jul 28 2018
===================

  * code style
  * update dependencies (epoll v2.0.3)

3.2.0 - Jul 24 2018
===================

  * add test to ensure HIGH and LOW have the expected values
  * add unittests (thank you [@pizzaisdavid](https://github.com/pizzaisdavid))
  * set active_low before setting direction in constructor
  * add constructor reconfigureDirection option

3.1.0 - May 13 2018
===================

  * replace new Buffer with Buffer.from or Buffer.alloc
  * add accessebile property to Gpio class (thank you [@johntalton](https://github.com/johntalton))
  * add HIGH and LOW properties to Gpio class (thank you [@johntalton](https://github.com/johntalton))

3.0.2 - Apr 07 2018
===================

  * update dependencies (epoll v2.0.1)
  * improve performance tests

3.0.1 - Apr 01 2018
===================

  * create poller for both inputs and outputs
  * add test to verify that gpio direction can be changed

3.0.0 - Mar 31 2018
===================

  * add effective debouncing support
  * codebase modernized
  * remove link to outdated tutorial
  * remove undocumented options method

2.0.0 - Feb 26 2018
===================

  * update dependencies (epoll v2.0.0)
  * drop support for node.js v0.10, v0.12, v5 and v7

1.2.0 - Feb 11 2018
===================

  * ignore edge argument when instantiating a Gpio for an output

1.1.9 - Dec 24 2017
===================

  * document node 9 support
  * update BeagleBone performance numbers
  * many documentation improvements
  * update BeagleBone Black performance numbers
  * update dependencies

1.1.8 - Oct 15 2017
===================

  * update dependencies (epoll v1.0.0)

1.1.7 - Aug 26 2017
===================

  * only check permissions for edge file if edge specified [#77](https://github.com/fivdi/onoff/issues/77)

1.1.5 - Jul 30 2017
===================

  * wait until unprivileged file access allowed

1.1.4 - Jul 15 2017
===================

  * improve examples

1.1.3 - Jun 18 2017
===================
  * upgrade to epoll v0.1.22
  * document related packages

1.1.2 - Feb 12 2017
===================
  * documentation improved
  * upgrade to epoll v0.1.21

1.1.1 - Jun 05 2016
===================
  * avoid exceptions when cape_universal is enabled on the bbb [#50](https://github.com/fivdi/onoff/issues/50)

1.1.0 - May 04 2016
===================
  * activeLow option
  * documentation improved

1.0.4 - Jan 29 2016
===================
  * documentation improved
  * epoll v0.1.17

1.0.3 - Oct 10 2015
===================
  * documentation improved
  * epoll v0.1.16

1.0.2 - Feb 18 2015
===================
  * documentation improved

1.0.1 - Feb 15 2015
===================
  * refactored tests to avoid relying in interrupt generating outputs as linux 3.13 and above no longer supports them
  * new wiring for tests and examples
  * pullup and pulldown resistor configuration documented

1.0.0 - Jan 10 2015
===================
  * use strict mode
  * jslint improvements

onoff.js
'use strict';

const fs = require('fs');
const debounce = require('lodash.debounce');
const Epoll = require('epoll').Epoll;

const GPIO_ROOT_PATH = '/sys/class/gpio/';

const HIGH_BUF = Buffer.from('1');
const LOW_BUF = Buffer.from('0');

const HIGH = 1;
const LOW = 0;

const exportGpio = gpio => {
  if (!fs.existsSync(gpio._gpioPath)) {
    // The GPIO hasn't been exported yet so export it
    fs.writeFileSync(GPIO_ROOT_PATH + 'export', '' + gpio._gpio);

    return false;
  }

  // The GPIO has already been exported, perhaps by onoff itself, perhaps
  // by quick2wire gpio-admin on the Pi, perhaps by the WiringPi gpio
  // utility on the Pi, or perhaps by something else. In any case, an
  // attempt is made to set the direction and edge to the requested
  // values here. If quick2wire gpio-admin was used for the export, the
  // user should have access to both direction and edge files. This is
  // important as gpio-admin sets niether direction nor edge. If the
  // WiringPi gpio utility was used, the user should have access to edge
  // file, but not the direction file. This is also ok as the WiringPi
  // gpio utility can set both direction and edge. If there are any
  // errors while attempting to perform the modifications, just keep on
  // truckin'.
  return true;
};

// Avoid the access permission issue described here:
// https://github.com/raspberrypi/linux/issues/553
// On some syetems udev rules are used to set access permissions on the GPIO
// sysfs files enabling those files to be accessed without root privileges.
// This takes a while so wait for it to complete.
const waitForGpioAccessPermission = (
  gpio, direction, edge, gpioPreviouslyExported
) => {
  let permissionRequiredPaths = [
    gpio._gpioPath + 'value',
  ];

  if (gpioPreviouslyExported === false) {
    permissionRequiredPaths.push(gpio._gpioPath + 'direction');
    permissionRequiredPaths.push(gpio._gpioPath + 'active_low');

    // On some systems the edge file will not exist if the GPIO does not
    // support interrupts
    // https://github.com/fivdi/onoff/issues/77#issuecomment-321980735
    if (edge && direction === 'in') {
      permissionRequiredPaths.push(gpio._gpioPath + 'edge');
    }
  }

  permissionRequiredPaths.forEach(path => {
    let tries = 0;

    while (true) {
      try {
        tries += 1;
        const fd = fs.openSync(path, 'r+');
        fs.closeSync(fd);
        break;
      } catch (e) {
        if (tries === 10000) {
          throw e;
        }
      }
    }
  });
};

const configureGpio = (
  gpio, direction, edge, options, gpioPreviouslyExported
) => {
  const throwIfNeeded = err => {
    if (gpioPreviouslyExported === false) {
      throw err;
    }
  };

  try {
    if (typeof options.activeLow === 'boolean') {
      gpio.setActiveLow(options.activeLow);
    }
  } catch (err) {
    throwIfNeeded(err);
  }

  try {
    const reconfigureDirection =
      typeof options.reconfigureDirection === 'boolean' ?
        options.reconfigureDirection : true;

    const requestedDirection =
      direction === 'high' || direction === 'low' ? 'out' : direction;

    if (reconfigureDirection || gpio.direction() !== requestedDirection) {
      gpio.setDirection(direction);
    }
  } catch (err) {
    throwIfNeeded(err);
  }

  try {
    // On some systems writing to the edge file for an output GPIO will
    // result in an "EIO, i/o error"
    // https://github.com/fivdi/onoff/issues/87
    if (edge && direction === 'in') {
      gpio.setEdge(edge);
    }
  } catch (err) {
    throwIfNeeded(err);
  }
};

const configureInterruptHandler = gpio => {
  // A poller is created for both inputs and outputs. A poller isn't
  // actually needed for an output but the setDirection method can be
  // invoked to change the direction of a GPIO from output to input and
  // then a poller may be needed.
  const pollerEventHandler = (err, fd, events) => {
    const value = gpio.readSync();

    if ((value === LOW && gpio._fallingEnabled) ||
        (value === HIGH && gpio._risingEnabled)) {
      gpio._listeners.slice(0).forEach(callback => {
        callback(err, value);
      });
    }
  };

  // Read GPIO value before polling to prevent an initial unauthentic
  // interrupt
  gpio.readSync();

  if (gpio._debounceTimeout > 0) {
    const db = debounce(pollerEventHandler, gpio._debounceTimeout);

    gpio._poller = new Epoll((err, fd, events) => {
      gpio.readSync(); // Clear interrupt
      db(err, fd, events);
    });
  } else {
    gpio._poller = new Epoll(pollerEventHandler);
  }
};

class Gpio {
  constructor(gpio, direction, edge, options) {
    if (typeof edge === 'object' && !options) {
      options = edge;
      edge = undefined;
    }

    options = options || {};

    this._gpio = gpio;
    this._gpioPath = GPIO_ROOT_PATH + 'gpio' + this._gpio + '/';
    this._debounceTimeout = options.debounceTimeout || 0;
    this._readBuffer = Buffer.alloc(16);
    this._readSyncBuffer = Buffer.alloc(16);
    this._listeners = [];

    const gpioPreviouslyExported = exportGpio(this);

    waitForGpioAccessPermission(
      this, direction, edge, gpioPreviouslyExported
    );

    configureGpio(this, direction, edge, options, gpioPreviouslyExported);

    this._valueFd = fs.openSync(this._gpioPath + 'value', 'r+');

    configureInterruptHandler(this);
  }

  read(callback) {
    const readValue = callback => {
      fs.read(this._valueFd, this._readBuffer, 0, 1, 0, (err, bytes, buf) => {
        if (typeof callback === 'function') {
          if (err) {
            return callback(err);
          }

          callback(null, convertBufferToBit(buf));
        }
      });
    };

    if (callback) {
      readValue(callback);
    } else {
      return new Promise((resolve, reject) => {
        readValue((err, value) => {
          if (err) {
            reject(err);
          } else {
            resolve(value);
          }
        });
      });
    }
  }

  readSync() {
    fs.readSync(this._valueFd, this._readSyncBuffer, 0, 1, 0);
    return convertBufferToBit(this._readSyncBuffer);
  }

  write(value, callback) {
    const writeValue = (value, callback) => {
      const writeBuffer = convertBitToBuffer(value);
      fs.write(
        this._valueFd, writeBuffer, 0, writeBuffer.length, 0, callback
      );
    };

    if (callback) {
      writeValue(value, callback);
    } else {
      return new Promise((resolve, reject) => {
        writeValue(value, err => {
          if (err) {
            reject(err);
          } else {
            resolve();
          }
        });
      });
    }
  }

  writeSync(value) {
    const writeBuffer = convertBitToBuffer(value);
    fs.writeSync(this._valueFd, writeBuffer, 0, writeBuffer.length, 0);
  }

  watch(callback) {
    this._listeners.push(callback);

    if (this._listeners.length === 1) {
      this._poller.add(this._valueFd, Epoll.EPOLLPRI);
    }
  }

  unwatch(callback) {
    if (this._listeners.length > 0) {
      if (typeof callback !== 'function') {
        this._listeners = [];
      } else {
        this._listeners = this._listeners.filter(listener => {
          return callback !== listener;
        });
      }

      if (this._listeners.length === 0) {
        this._poller.remove(this._valueFd);
      }
    }
  }

  unwatchAll() {
    this.unwatch();
  }

  direction() {
    return fs.readFileSync(this._gpioPath + 'direction').toString().trim();
  }

  setDirection(direction) {
    fs.writeFileSync(this._gpioPath + 'direction', direction);
  }

  edge() {
    return fs.readFileSync(this._gpioPath + 'edge').toString().trim();
  }

  setEdge(edge) {
    fs.writeFileSync(this._gpioPath + 'edge', edge);

    this._risingEnabled = edge === 'both' || edge === 'rising';
    this._fallingEnabled = edge === 'both' || edge === 'falling';
  }

  activeLow() {
    return convertBufferToBoolean(
      fs.readFileSync(this._gpioPath + 'active_low')
    );
  }

  setActiveLow(invert) {
    fs.writeFileSync(
      this._gpioPath + 'active_low', convertBooleanToBuffer(!!invert)
    );
  }

  unexport() {
    this.unwatchAll();
    fs.closeSync(this._valueFd);
    try {
      fs.writeFileSync(GPIO_ROOT_PATH + 'unexport', '' + this._gpio);
    } catch (ignore) {
      // Flow of control always arrives here when cape_universal is enabled on
      // the bbb.
    }
  }

  static get accessible() {
    let fd;

    try {
      fd = fs.openSync(GPIO_ROOT_PATH + 'export', fs.constants.O_WRONLY);
    } catch(e) {
      // e.code === 'ENOENT' / 'EACCES' are most common
      // though any failure to open will also result in a gpio
      // failure to export.
      return false;
    } finally {
      if (fd) {
        fs.closeSync(fd);
      }
    }

    return true;
  }
}

const convertBitToBuffer = bit => convertBooleanToBuffer(bit === HIGH);
const convertBufferToBit =
  buffer => convertBufferToBoolean(buffer) ? HIGH : LOW;

const convertBooleanToBuffer = boolean => boolean ? HIGH_BUF : LOW_BUF;
const convertBufferToBoolean = buffer => buffer[0] === HIGH_BUF[0];

Gpio.HIGH = HIGH;
Gpio.LOW = LOW;

module.exports.Gpio = Gpio;

